### ISSUE TITLE
atomic_compare_exchange undefined for pair_int_t type

### ISSUE THREAD

**Issue by j8asic** [kokkos/kokkos#3344]
I'm compiling Kokkos with `clang++` on `Windows` for `CUDA` backend (not `clang-cl`, since it doesn't seem to pass CUDA flags). Windows atomics is automatically chosen:
```
#if defined(_WIN32)
#define KOKKOS_ENABLE_WINDOWS_ATOMICS
```
But then compilation fails with missing definitions of atomic functions. And then if I add a test to check for using `clang`, and in such case define `KOKKOS_ENABLE_GNU_ATOMICS`, _everything goes okay_, except for `HostThreadTeamData::get_work_stealing()`, which calls `Kokkos::atomic_compare_exchange` using `pair_int_t` arguments. This is the output:
```
[10/23] Building CXX object core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_HostThreadTeam.cpp.obj
FAILED: core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_HostThreadTeam.cpp.obj 
C:\PROGRA~1\LLVM\bin\CLANG_~1.EXE -D_CRT_SECURE_NO_WARNINGS -Icore/src -I../core/src -I. -isystem "C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v9.2/include" --cuda-path="
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v9.2" -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -mcx16 -x cuda --cuda-gpu-arch=sm_52 -std=c++17 -MD -MT core/sr
c/CMakeFiles/kokkoscore.dir/impl/Kokkos_HostThreadTeam.cpp.obj -MF core\src\CMakeFiles\kokkoscore.dir\impl\Kokkos_HostThreadTeam.cpp.obj.d -o core/src/CMakeFiles/kokkoscore.dir/
impl/Kokkos_HostThreadTeam.cpp.obj -c ../core/src/impl/Kokkos_HostThreadTeam.cpp
../core/src/impl/Kokkos_HostThreadTeam.cpp:224:11: error: call to 'atomic_compare_exchange' is ambiguous
      w = Kokkos::atomic_compare_exchange(&m_work_range, w, w_new);
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../core/src/impl/Kokkos_Atomic_Windows.hpp:139:10: note: candidate function [with T = Kokkos::pair<long long, long long>]
inline T atomic_compare_exchange(
         ^
../core/src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp:253:10: note: candidate function [with T = Kokkos::pair<long long, long long>]
inline T atomic_compare_exchange(
         ^
../core/src/impl/Kokkos_HostThreadTeam.cpp:257:13: error: call to 'atomic_compare_exchange' is ambiguous
        w = Kokkos::atomic_compare_exchange(steal_range, w, w_new);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../core/src/impl/Kokkos_Atomic_Windows.hpp:139:10: note: candidate function [with T = Kokkos::pair<long long, long long>]
inline T atomic_compare_exchange(
         ^
../core/src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp:253:10: note: candidate function [with T = Kokkos::pair<long long, long long>]
inline T atomic_compare_exchange(
         ^
2 errors generated when compiling for host.
```

What do you recommend to solve this issue? This mixing of MSVC and clang defines on Windows is totally chaotic...

**Comment by j8asic** [comment]
Here are the ideas:  

Option 1: As in above post, use Clang's GNU atomics. Allow that ` _atomic_compare_exchange_strong` accepts `sizeof(T) == 16`.  

Option 2: Leave usage of Windows atomics. Comment out `#ifndef __CUDA_ARCH__` in `Kokkos_Atomic_Windows.hpp` to enable `atomic_compare_exchange` (anyway why host fns need to be hidden if Cuda compiler is active?). Compiler now gives an error there is no `atomic_fetch_add` definition for host. Solution: implement it in `Kokkos_Atomic_Windows.hpp` using `_InterlockedExchangeAdd***`?

**Comment by j8asic** [comment]
Implemented a mix of two options in #3345: left Win atomics as above, but also enabled GNU atomics when CUDA is enabled. But it would be more clean if `atomic_fetch_add` is also implemented for Win...
Will close the issue when PR gets merged.
